airtable/airtable_to_mailing_list.py

```diff
@@ -0,0 +1,136 @@
+    and returns a dictionary that maps chapter ids to mailing list
+    emails.
+    """
+    chapter_to_mailing_list = {}
+    email_col_name = "Mailing List Email"
+    for c in chapters:
+        if "fields" in c and email_col_name in c["fields"]:
+            email = c["fields"][email_col_name]
+            # Sanity check that the email is probably valid.
+            if "@" not in email:
+                raise Exception("Mailing list email for chapter {} is malformed: {}".format(c, email))
+            c_id = c["id"]
+            chapter_to_mailing_list[c_id] = email
+    return chapter_to_mailing_list
+
+
+# AddMember is a namedtuple for airtable members.
+#  - "airtable_id" is the member's airtable id
+#  - "email" is the member's email
+#  - "chapters_added" is a list of chapter ids where the member has been
+# added to the chapter mailing list
+#  - "chapters_to_add" is a list of chapter ids that represents the
+# chapters that the member hasn't been added to the mailing list for.
+AddMember = namedtuple("AddMember", ["airtable_id", "email", "chapters_added", "chapters_to_add"])
+
+CHAPTER_ID_COL = "chapter_id"
+OPTED_OUT_COL = "Opted Out Of Communication"
+ADDED_TO_MAILING_LIST_COL = "(DO NOT MODIFY) Added To Mailing List"
+
+def get_members_to_add(members):
+    """Filters out members that do not need to be added to a mailing list.
+
+    Filters a list of airtable members that have the form:
+
+    {
+    "id": string,
+    "fields": {
+              "email": optional string,
+              "chapter_id": optional list of strings,
+              "(DO NOT MODIFY) Added To Mailing List": optional list of strings,
+              "Opted Out Of Communication": optional bool
+             }
+    }
+
+    and filters out the members that have 1) opted out of communication,
+    and 2) do not need to be added to a mailing list. Returns a list of
+    AddMembers that need to be added to mailing lists.
+    """
+    members_to_add = []
+    for m in members:
+        if "fields" not in m:
+            continue # m has no fields.
+        fields = m["fields"]
+        # Skip m if they opted out of communication.
+        if fields.get(OPTED_OUT_COL, False):
+            continue
+        # Skip m if they do not have an email.
+        if "email" not in fields:
+            continue
+        chapters = set(fields.get(CHAPTER_ID_COL, []))
+        chapters_added = set(fields.get(ADDED_TO_MAILING_LIST_COL, []))
+        chapters_to_add = chapters - chapters_added
+        if len(chapters_to_add) == 0:
+            continue # m is in all the mailing lists that they belong to.
+        members_to_add.append(AddMember(airtable_id=m["id"],
+                                        email=fields["email"],
+                                        chapters_added=chapters_added,
+                                        chapters_to_add=chapters_to_add))
+    return members_to_add
+
+
+def main():
+    # chapters is all of the chapters in the "Chapters" table.
```

Comment:
rename this, or factor most of it out into another function, or add a docstring explaining what this function does